File: 59. Spiral Matrix II/59. Spiral Matrix II.rb
# @param {Integer} n
# @return {Integer[][]}
def generate_matrix(n)
    matrix = Array.new(n) { Array.new(n, 0) }
    left, right, top, bottom = 0, n - 1, 0, n - 1
    num = 1
    
    while left <= right && top <= bottom
        (left..right).each do |i|
            matrix[top][i] = num
            num += 1
        end
        top += 1
        
        (top..bottom).each do |i|
            matrix[i][right] = num
            num += 1
        end
        right -= 1
        
        if top <= bottom
            right.downto(left).each do |i|
                matrix[bottom][i] = num
                num += 1
            end
            bottom -= 1
        end
        
        if left <= right
            bottom.downto(top).each do |i|
                matrix[i][left] = num
                num += 1
            end
            left += 1
        end
    end
    
    matrix
end

File: 59. Spiral Matrix II/59. Spiral Matrix II.js
function generateMatrix(n) {
    // Initialize the matrix with zeros
    const matrix = Array.from({ length: n }, () => Array(n).fill(0));

    let num = 1;
    let top = 0;
    let bottom = n - 1;
    let left = 0;
    let right = n - 1;

    while (top <= bottom && left <= right) {
        // Traverse from left to right
        for (let i = left; i <= right; i++) {
            matrix[top][i] = num++;
        }
        top++;

        // Traverse from top to bottom
        for (let i = top; i <= bottom; i++) {
            matrix[i][right] = num++;
        }
        right--;

        // Traverse from right to left (if still within bounds)
        if (top <= bottom) {
            for (let i = right; i >= left; i--) {
                matrix[bottom][i] = num++;
            }
            bottom--;
        }

        // Traverse from bottom to top (if still within bounds)
        if (left <= right) {
            for (let i = bottom; i >= top; i--) {
                matrix[i][left] = num++;
            }
            left++;
        }
    }

    return matrix;
}

// Example usage:
console.log(generateMatrix(3));
// Output: [[1,2,3],[8,9,4],[7,6,5]]

console.log(generateMatrix(1));
// Output: [[1]]